crates/astria-conductor/src/executor/channel.rs

```diff
@@ -0,0 +1,194 @@
+    #[error("no permits available")]
+    NoPermits(T),
+}
+
+impl<T> TrySendError<T> {
+    fn from_semaphore(err: &TryAcquireError, block: T) -> Self {
+        match err {
+            tokio::sync::TryAcquireError::Closed => Self::Closed(block),
+            tokio::sync::TryAcquireError::NoPermits => Self::NoPermits(block),
+        }
+    }
+}
+
+impl<T> From<TokioSendError<T>> for TrySendError<T> {
+    fn from(err: TokioSendError<T>) -> Self {
+        Self::Closed(err.0)
+    }
+}
+
+#[derive(Debug, Clone)]
+pub(super) struct Sender<T> {
+    sem: Arc<Semaphore>,
+    chan: UnboundedSender<T>,
+}
+
+impl<T> Sender<T> {
+    /// Sends a block, waiting until the channel has permits.
+    ///
+    /// Returns an error if the channel is closed.
+    pub(super) async fn send(&self, block: T) -> Result<(), SendError> {
+        let permit = self.sem.acquire().await?;
+        permit.forget();
+        self.chan.send(block)?;
+        Ok(())
+    }
+
+    /// Attempts to send a block without blocking.
+    ///
+    /// Returns an error if the channel is out of permits or if it has been closed.
+    // allow: this is mimicking tokio's `SendError` that returns the stack-allocated object.
+    #[allow(clippy::result_large_err)]
+    pub(super) fn try_send(&self, block: T) -> Result<(), TrySendError<T>> {
+        let permit = match self.sem.try_acquire() {
+            Ok(permit) => permit,
+            Err(err) => return Err(TrySendError::from_semaphore(&err, block)),
+        };
+        permit.forget();
+        self.chan.send(block)?;
+        Ok(())
+    }
+}
+
+pub(super) struct Receiver<T> {
+    cap: usize,
+    sem: Arc<Semaphore>,
+    chan: UnboundedReceiver<T>,
+}
+
+impl<T> Receiver<T> {
+    /// Sets the channel's capacity to `cap`.
+    ///
+    /// `cap` will be the maximum number of blocks that can be sent
+    /// over the channel before new permits are added with `[SoftBlockReceiver::add_permits]`.
+    pub(super) fn set_capacity(&mut self, cap: usize) {
+        self.cap = cap;
+    }
+
+    /// Adds up to `capacity` number of permits to the channel.
+    ///
+    /// `capacity` is previously set by [`SoftBlockReceiver::set_capacity`]
+    /// or zero by default.
+    pub(super) fn fill_permits(&self) {
+        let additional = self.cap.saturating_sub(self.sem.available_permits());
+        self.sem.add_permits(additional);
+    }
+
+    /// Receives a block over the channel.
+    pub(super) async fn recv(&mut self) -> Option<T> {
+        self.chan.recv().await
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::{
+        soft_block_channel,
+        TrySendError,
+    };
+
+    #[test]
+    fn fresh_channel_has_no_capacity() {
+        let (tx, _rx) = soft_block_channel::<()>();
+        assert_eq!(
+            tx.try_send(()).unwrap_err(),
+            TrySendError::NoPermits(()),
+            "a fresh channel starts without permits"
+        );
+    }
+
+    #[test]
+    fn permits_are_filled_to_capacity() {
```

Comment:
Should this test first expect 1 to be filled, and fail only when at capacity?

Reply:
I am not sure I understand. Can you give me an example?

Reply:
Nevermind, misread the test.